test/worker_loop_test.dart

```diff
@@ -2,32 +2,57 @@
 // for details. All rights reserved. Use of this source code is governed by a
 // BSD-style license that can be found in the LICENSE file.
 
+import 'dart:async';
+import 'dart:io';
+
 import 'package:test/test.dart';
 
 import 'package:bazel_worker/bazel_worker.dart';
 import 'package:bazel_worker/testing.dart';
 
 void main() {
+  group('SyncWorkerLoop', () {
+    runTests(
+        (Stdin stdinStream, Stdout stdoutStream) =>
+            new TestSyncWorkerConnection(stdinStream, stdoutStream),
+        (TestSyncWorkerConnection connection) =>
+            new TestSyncWorkerLoop(connection));
+  });
+
+  group('AsyncWorkerLoop', () {
+    runTests(
+        (Stream<List<int>> stdinStream, StreamSink<List<int>> stdoutStream) =>
+            new TestAsyncWorkerConnection(stdinStream, stdoutStream),
+        (TestAsyncWorkerConnection connection) =>
+            new TestAsyncWorkerLoop(connection));
+  });
+}
+
+void runTests(
+    workerConnectionFactory(stdin, stdout), workerLoopFactory(connection)) {
   TestStdinStream stdinStream;
   TestStdoutStream stdoutStream;
-  TestSyncWorkerConnection connection;
-  TestSyncWorkerLoop workerLoop;
+  // TestSyncWorkerConnection or TestAsyncWorkerConnection
+  var connection;
```

Comment:
Consider making a common base class for TestSyncWorkerConnection and TestAsyncWorkerConnection, so that we can express precisely what type of object appears here.